modules/catalog-compaction/src/main/java/org/apache/ignite/internal/catalog/compaction/CatalogCompactionRunner.java

```diff
@@ -401,74 +441,94 @@ private static List<String> missingNodes(Set<String> requiredNodes, Collection<L
+    }
+
+    CompletableFuture<Boolean> propagateTimeToReplicasLocal(long txBeginTime) {
         HybridTimestamp nowTs = clockService.now();
 
         return schemaSyncService.waitForMetadataCompleteness(nowTs)
                 .thenComposeAsync(ignore -> {
-                    Map<Integer, Integer> tablesWithPartitions =
-                            catalogManagerFacade.collectTablesWithPartitionsBetween(timestamp, nowTs.longValue());
+                    AtomicBoolean abortFlag = new AtomicBoolean();
+                    Int2IntMap tablesWithPartitions =
+                            catalogManagerFacade.collectTablesWithPartitionsBetween(txBeginTime, nowTs.longValue());
 
-                    Set<String> topologyNodeNames = topologyNodes.stream()
-                            .map(ClusterNode::name)
-                            .collect(Collectors.toSet());
+                    ObjectIterator<Entry> itr = tablesWithPartitions.int2IntEntrySet().iterator();
 
-                    // TODO https://issues.apache.org/jira/browse/IGNITE-22951 Minimize the number of network requests
-                    return CompletableFutures.allOf(tablesWithPartitions.entrySet().stream()
-                            .map(e -> invokeOnReplicas(e.getKey(), e.getValue(), timestamp, nowTs, topologyNodeNames))
-                            .collect(Collectors.toList())
-                    );
+                    return invokeOnLocalReplicas(txBeginTime, itr, abortFlag)
+                            .thenApply(v -> !abortFlag.get());
                 }, executor);
     }
 
-    private CompletableFuture<Void> invokeOnReplicas(
-            int tableId,
-            int partitions,
-            long txBeginTime,
-            HybridTimestamp nowTs,
-            Set<String> logicalTopologyNodes
-    ) {
-        List<TablePartitionId> replicationGroupIds = new ArrayList<>(partitions);
+    private CompletableFuture<Void> invokeOnLocalReplicas(long txBeginTime, ObjectIterator<Entry> tabTtr, AtomicBoolean abortFlag) {
+        if (!tabTtr.hasNext() || abortFlag.get()) {
+            return CompletableFutures.nullCompletedFuture();
+        }
+
+        Entry tableWithPartitions = tabTtr.next();
+        int tableId = tableWithPartitions.getIntKey();
+        int partitions = tableWithPartitions.getIntValue();
+        List<CompletableFuture<?>> partFutures = new ArrayList<>(partitions);
 
         for (int p = 0; p < partitions; p++) {
-            replicationGroupIds.add(new TablePartitionId(tableId, p));
-        }
+            TablePartitionId tablePartitionId = new TablePartitionId(tableId, p);
 
-        return placementDriver.getAssignments(replicationGroupIds, nowTs)
-                .thenComposeAsync(tokenizedAssignments -> {
-                    assert tokenizedAssignments.size() == replicationGroupIds.size();
+            CompletableFuture<?> fut = placementDriver
+                    .getPrimaryReplica(tablePartitionId, clockService.now())
```

Comment:
does it make sense to acquire time once and use it for every partition?

Reply:
Done, thanks.